src/runtime/security/kms_client.cpp

```diff
@@ -0,0 +1,204 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <initializer_list>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include "absl/strings/escaping.h"
+#include "fmt/core.h"
+#include "http/http_client.h"
+#include "http/http_method.h"
+#include "http/http_status_code.h"
+#include "nlohmann/json.hpp"
+#include "nlohmann/json_fwd.hpp"
+#include "runtime/security/kms_client.h"
+#include "runtime/security/replica_kms_info.h"
+#include "utils/error_code.h"
+#include "utils/fmt_logging.h"
+
+namespace dsn {
+namespace security {
+
+dsn::error_s KMSClient::DecryptEncryptionKey(const dsn::replication::replica_kms_info &kms_info,
+                                             std::string *decrypted_key)
+{
+    nlohmann::json payload;
+    payload["name"] = cluster_key_name_;
+    std::string iv_plain = ::absl::HexStringToBytes(kms_info.iv);
+    std::string iv_b64;
+    ::absl::WebSafeBase64Escape(iv_plain, &iv_b64);
+    payload["iv"] = iv_b64;
+    std::string eek_plain = ::absl::HexStringToBytes(kms_info.eek);
+    std::string eek_b64;
+    ::absl::WebSafeBase64Escape(eek_plain, &eek_b64);
+    payload["material"] = eek_b64;
+
+    http_client client;
+    auto err = client.init();
+    if (!err.is_ok()) {
+        return dsn::error_s::make(ERR_CURL_FAILED, "Start http client failed");
+    }
+
+    err = client.set_auth(http_auth_type::SPNEGO);
+    if (!err.is_ok()) {
+        return dsn::error_s::make(ERR_CURL_FAILED, "http client set auth type failed");
+    }
+
+    std::vector<std::string> urls;
+    urls.reserve(kms_urls_.size());
+    for (const auto &url : kms_urls_) {
+        urls.emplace_back(fmt::format("{}/v1/keyversion/{}/_eek?eek_op=decrypt", url, kms_info.kv));
+    }
+
+    client.clear_header_fields();
+    client.set_content_type("application/json");
+    client.set_accept("*/*");
+    err = client.with_post_method(payload.dump());
+    if (!err.is_ok()) {
+        return dsn::error_s::make(ERR_CURL_FAILED, "http client set method failed");
+    }
+
+    nlohmann::json j;
+    for (const auto &url : urls) {
+        err = client.set_url(url);
+        if (!err.is_ok()) {
+            return dsn::error_s::make(ERR_CURL_FAILED, "http clientt set url failed");
+        }
+        std::string resp;
+        err = client.exec_method(&resp);
+        if (!err.is_ok()) {
+            return dsn::error_s::make(ERR_CURL_FAILED, "http client exec post method failed");
```

Comment:
In the case of one of the KMS servers down, exec_method will return error, so it should contine to try on the next KMS server. In Kudu, it continue in the case of 

```
if (s.IsNetworkError() || s.IsTimedOut()) {
  continue;
}
```